kafka/src/main/java/io/micronaut/configuration/kafka/intercept/KafkaClientIntroductionAdvice.java

```diff
@@ -475,14 +478,28 @@ private MessagingClientException wrapException(MethodInvocationContext<Object, O
     }
 
     private ProducerRecord<?, ?> buildProducerRecord(MethodInvocationContext<Object, Object> context, ProducerState producerState, Object value) {
-        return new ProducerRecord<>(
+        return decorateProducerRecord(context, producerState, value)
+            .orElseGet(() -> new ProducerRecord<>(
                 producerState.topicSupplier.get(context),
                 producerState.partitionSupplier.get(context),
                 producerState.timestampSupplier.get(context),
                 producerState.keySupplier.get(context),
                 value,
-                producerState.headersSupplier.get(context)
-        );
+                producerState.headersSupplier.get(context)));
+    }
+
+    private Optional<ProducerRecord<?, ?>> decorateProducerRecord(MethodInvocationContext<Object, Object> context, ProducerState producerState, Object value) {
+        return Optional.ofNullable(value)
+            .filter(ProducerRecord.class::isInstance).map(ProducerRecord.class::cast)
+            .map(r -> new ProducerRecord<>(
+                Optional.ofNullable(r.topic()).filter(not(String::isEmpty)).orElseGet(() -> producerState.topicSupplier.get(context)),
+                Optional.ofNullable(r.partition()).orElseGet(() -> producerState.partitionSupplier.get(context)),
+                Optional.ofNullable(r.timestamp()).orElseGet(() -> producerState.timestampSupplier.get(context)),
+                Optional.ofNullable(r.key()).orElseGet(() -> producerState.keySupplier.get(context)),
+                r.value(),
+                concat(Optional.ofNullable(producerState.headersSupplier.get(context)).stream(),
+                    Optional.ofNullable(r.headers()).map(Headers::toArray).map(Arrays::asList).stream())
```

Comment:
I think here you can do `flatMap(Arrays::Stream)` directly

Reply:
Thanks for your comment @dstepanov 

I don't think that would work in this case though. You see, I need to to `concat` two compatible streams here:

- The first one (`headersSupplier` in line 500) is a `Stream<Collection<Header>>`.
- The second one (`r.headers` in line 501) is a `Stream<List<Header>>`.

Right after that, the concatenated streams are flat-mapped with `Collection::stream` then collected `toList` (line 502).

So if we were to use `Arrays::Stream`, the simplest expression I can think of is:

```java
concat(
    Optional.ofNullable(producerState.headersSupplier.get(context)).map(Collection::stream).orElseGet(Stream::empty),
    Optional.ofNullable(r.headers()).map(Headers::toArray).map(Arrays::stream).orElseGet(Stream::empty)
).toList()
```

which I don't think is particularly more readable than the way it's implemented now:

```java
concat(
    Optional.ofNullable(producerState.headersSupplier.get(context)).stream(),
    Optional.ofNullable(r.headers()).map(Headers::toArray).map(Arrays::asList).stream()
).flatMap(Collection::stream).toList()
```

But I may be wrong 🤔  Do you have a simpler way to merge those two collections of Headers together?

Reply:
It's not that important :)
I would write:
```
                        Optional.ofNullable(producerState.headersSupplier.get(context)).stream().flatMap(Collection::stream),
                        Optional.ofNullable(r.headers()).stream().flatMap(h -> Arrays.stream(h.toArray())))     
```